Source/WebCore/Modules/permissions/Permissions.cpp

```diff
@@ -75,62 +80,74 @@ static bool isAllowedByFeaturePolicy(const Document& document, PermissionName na
-        return;
-    }
+    if (document) {
+        if (!document->isFullyActive()) {
+            promise.reject(Exception { InvalidStateError, "The document is not fully active"_s });
+            return;
+        }
 
-    if (!permissionDescriptorValue) {
-        promise.reject(Exception { DataError, "The parameter is invalid"_s });
-        return;
+        if (!document->page()) {
+            promise.reject(Exception { NotSupportedError });
+            return;
+        }
     }
 
-    JSC::VM& vm = context->globalObject()->vm();
+    JSC::VM& vm = context.globalObject()->vm();
     auto scope = DECLARE_THROW_SCOPE(vm);
-    auto parameterDescriptor = convert<IDLDictionary<PermissionDescriptor>>(*context->globalObject(), permissionDescriptorValue.get());
+    auto parameterDescriptor = convert<IDLDictionary<PermissionDescriptor>>(*context.globalObject(), permissionDescriptorValue.get());
     if (UNLIKELY(scope.exception())) {
         promise.reject(Exception { ExistingExceptionError });
         return;
     }
 
-    if (is<Document>(context) && !isAllowedByFeaturePolicy(downcast<Document>(*context), parameterDescriptor.name)) {
-        context->postTask([parameterDescriptor, promise = WTFMove(promise)](auto& context) mutable {
-            promise.resolve(PermissionStatus::create(context, PermissionState::Denied, parameterDescriptor));
-        });
-        return;
+    if (is<Document>(context)) {
+        if (!isAllowedByFeaturePolicy(downcast<Document>(context), parameterDescriptor.name)) {
+            context.postTask([parameterDescriptor, promise = WTFMove(promise)](auto& context) mutable {
+                promise.resolve(PermissionStatus::create(context, PermissionState::Denied, parameterDescriptor));
+            });
+            return;
+        }
     }
 
-    auto* origin = context->securityOrigin();
+    auto* origin = context.securityOrigin();
     auto originData = origin ? origin->data() : SecurityOriginData { };
 
-    PermissionController::shared().query(ClientOrigin { context->topOrigin().data(), originData }, PermissionDescriptor { parameterDescriptor }, *document->page(), [this, protectedThis = Ref { *this }, parameterDescriptor, promise = WTFMove(promise)](auto permissionState) mutable {
-        auto context = m_navigator ? m_navigator->scriptExecutionContext() : nullptr;
-        if (!context || !context->globalObject())
+    if (is<WorkerGlobalScope>(context)) {
+        auto* workerGlobalScope = downcast<WorkerGlobalScope>(&context);
+
+        auto completionHandler = [originData, parameterDescriptor, identifier = workerGlobalScope->identifier(), promise = WTFMove(promise)] (auto& context) mutable { 
+            PermissionController::shared().query(ClientOrigin { context.topOrigin().data(), originData }, PermissionDescriptor { parameterDescriptor }, *downcast<Document>(context).page(), [&context, identifier, parameterDescriptor, promise = WTFMove(promise)](auto permissionState) mutable {
```

Comment:
I don't think it is safe to assume that `downcast<Document>(context).page()` is non-null here.

Reply:
You cannot capture context by reference since the lambda will run asynchronously. You'll have a Ref the context and capture the Ref<>.

Reply:
WTFMove(originData)

Reply:
Fixed!